server/src/main/java/org/elasticsearch/common/geo/QuadkeyUtils.java

```diff
@@ -0,0 +1,189 @@
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.support.XContentMapValues;
+
+import java.io.IOException;
+
+import static org.elasticsearch.common.geo.GeoUtils.normalizeLat;
+import static org.elasticsearch.common.geo.GeoUtils.normalizeLon;
+
+/**
+ * Implements quad key hashing, same as used by map tiles.
+ * The string key is formatted as  "zoom/x/y"
+ * The hash value (long) contains all three of those values.
+ */
+public class QuadkeyUtils {
+
+    /**
+     * Largest number of tiles (precision) to use.
+     * This value cannot be more than (64-5)/2 = 29, because 5 bits are used for zoom level itself
+     * If zoom is not stored inside hash, it would be possible to use up to 32.
+     * Another consideration is that index optimizes lat/lng storage, loosing some precision.
+     * E.g. hash lng=140.74779717298918D lat=45.61884022447444D == "18/233561/93659", but shown as "18/233561/93658"
+     */
+    public static final int MAX_ZOOM = 29;
+
+    /**
+     * Bit position of the zoom value within hash.  Must be &gt;= 2*MAX_ZOOM
+     * Keeping it at a constant place allows MAX_ZOOM to be increased
+     * without breaking serialization binary compatibility
+     * (still, the newer version should not use higher MAX_ZOOM in the mixed cases)
+     */
+    private static final int ZOOM_SHIFT = 29 * 2;
+
+    /**
+     * Mask of all the bits used by the quadkey in a hash
+     */
+    private static final long QUADKEY_MASK = (1L << ZOOM_SHIFT) - 1;
+
+    /**
+     * Parse quadkey hash as zoom, x, y integers.
+     */
+    private static int[] parseHash(final long hash) {
+        final int zoom = checkPrecisionRange((int) (hash >>> ZOOM_SHIFT));
+        final int tiles = 1 << zoom;
+
+        // decode the quadkey bits as interleaved xTile and yTile
+        long val = hash & QUADKEY_MASK;
+        int xTile = (int) BitUtil.deinterleave(val);
+        int yTile = (int) BitUtil.deinterleave(val >>> 1);
+        if (xTile < 0 || yTile < 0 || xTile >= tiles || yTile >= tiles) {
+            throw new IllegalArgumentException("hash-tile");
+        }
+
+        return new int[]{zoom, xTile, yTile};
+    }
+
+    private static double tile2lon(final double x, final double tiles) {
+        return x / tiles * 360.0 - 180;
+    }
+
+    private static double tile2lat(final double y, final double tiles) {
+        double n = Math.PI - (2.0 * Math.PI * y) / tiles;
+        return Math.toDegrees(Math.atan(Math.sinh(n)));
+    }
+
+    /**
+     * Parse a precision that can be expressed as an integer or a distance measure like "1km", "10m".
+     *
+     * The precision is expressed as a zoom level between 0 and MAX_ZOOM.
+     *
+     * @param parser {@link XContentParser} to parse the value from
+     * @return int representing precision
+     */
+    public static int parsePrecision(XContentParser parser) throws IOException, ElasticsearchParseException {
```

Comment:
Two questions:

1. Should we support unit-less precision levels?  It looks like it adds a lot of autodetect magic below, and we've consistently been moving towards requiring units everywhere in ES.  I don't know enough about the algo to say, but is there something tangible that we gain by supporting unit-less precision over just `10km`, etc?
2. Is there a reason we fall back to the "old" style of xcontent parsing instead of using the newer static parsers here too?

Reply:
1) for quadkey, most of the time it will be unitless - e.g .map visualizations will use `current_zoom + 2` as the precision to get 16 dots per visualization tile.
2) I tried to make as few changes as possible to the existing code to make reviewing easier. This code is an exact copy of the `GeoUtils.parsePrecision()`, adjusted for different validation and parsing. We might consider rewriting them, but I think we should do them at the same time and possibly consolidate it for reuse.

Reply:
@polyfractal, per @talevy 's suggestion, I removed non-integer precision - it is not needed by our use cases, and we do not know if they will be needed by other users. This also simplifies precision parsing.